### PR TITLE
Items: Fix creation of "default" NetworkInterface

### PR DESCRIPTION
**Pull request by btravouillon**
Do not force creation of a network interface named `default`.

Closes: #2838

### CODE REVIEW

**Comment by SchoolGuy** [comment]
@btravouillon The approach is not so bad to be honest. What we now need is to ensure that the code doesn't access the network interfaces. I tried to play it safe and always have one (because DHCP and DNS configuration is hard without a MAC for example) but sure we can make the network interfaces optional.

The tests are mainly failing because of my hardcoded name for the interfaces in there (`default`). If we make `__get_interface` public, add a warning into the log when an interface is implicitly created and then use that function in the tests, then I guess we should be good already according to the tests.

**Comment by btravouillon** [comment]
I need to do additional testing on my side, I still get the `default` interface created after `sudo cobbler system edit --name cobbler2838 --netboot 1`.
I wonder if this could be related to https://github.com/cobbler/cobbler/blob/v3.3.3/cobbler/remote.py#L2108.

For the tests, my approach is to explicitely declare the `default` interface whenever needed.

For example:
```
diff --git a/tests/items/system_test.py b/tests/items/system_test.py
index 8510b2fe..c58b69a4 100644
--- a/tests/items/system_test.py
+++ b/tests/items/system_test.py
@@ -658,6 +658,7 @@ def test_serial_baud_rate(cobbler_api, value, expected_exception):
 def test_from_dict_with_network_interface(cobbler_api):
     # Arrange
     system = System(cobbler_api)
+    system.interfaces = { "default": NetworkInterface(cobbler_api) }
     sys_dict = system.to_dict()
 
     # Act
@@ -682,6 +683,7 @@ def test_is_management_supported(
 ):
     # Arrange
     system = System(cobbler_api)
+    system.interfaces = { "default": NetworkInterface(cobbler_api) }
     system.interfaces["default"].mac_address = input_mac
     system.interfaces["default"].ip_address = input_ipv4
     system.interfaces["default"].ipv6_address = input_ipv6
```

Does that seem correct?

**Comment by btravouillon** [comment]
https://github.com/cobbler/cobbler/blob/dd45e6e80fbd2deba94ad3b719b7335164b7296e/cobbler/remote.py#L2257-L2259

I guess the **FIXME** is right :sweat_smile:.

**Comment by btravouillon** [comment]
**Linting / lint_docs** is faling while it did not yesterday: https://github.com/cobbler/cobbler/runs/6735620240?check_suite_focus=true

rstcheck 6.0.0 has been released since: https://github.com/rstcheck/rstcheck/releases/tag/v6.0.0

**Comment by SchoolGuy** [comment]
@btravouillon Your approach for the tests is fully correct imho.

**Comment by SchoolGuy** [comment]
@btravouillon In regard to the fixme... Yes sadly I wasn't able to remove that method. It is only used by the CLI atm. However the new CLI is not yet finished, so we need to fix that fixme.

**Comment by btravouillon** [comment]
> @btravouillon In regard to the fixme... Yes sadly I wasn't able to remove that method. It is only used by the CLI atm. However the new CLI is not yet finished, so we need to fix that fixme.

I have some code pending for push. Will do once back home.

**Comment by btravouillon** [comment]
I just noticed the `--interface` is not mandatory as per https://cobbler.readthedocs.io/en/latest/cobbler.html#cobbler-system. I've always believed this was mandatory...

Instead of `default`, we could consider the first network interface defined. However, this would need an ordered dict.

What is the minimum version of Python 3 required by Cobbler?

**Comment by SchoolGuy** [comment]
@btravouillon This is the fun part... I want to stay on 3.6 since there are distros still supporting that.

**Comment by SchoolGuy** [comment]
Idea: Let's make `--interface` obligatory in case we edit Interface specific things. I like explicit behaviour. That way we can easily fix the fixme I guess?

**Comment by btravouillon** [comment]
> Idea: Let's make `--interface` obligatory in case we edit Interface specific things. I like explicit behaviour. That way we can easily fix the fixme I guess?

I plan to replace this code with the logic below in `CobblerXMLRPCInterface.__interface_edits()`:
https://github.com/cobbler/cobbler/blob/dd45e6e80fbd2deba94ad3b719b7335164b7296e/cobbler/remote.py#L2296-L2297

- if `--interface foo` is passed as an argument, use `foo` (current behavior)
- if the system has no interface and `--interface` is not defined, use `default` (current behavior)
- if the system has a single interface and `--interface` is not defined, use this interface
- if the system has several interfaces, `--interface` is not defined, and `default` exists, use `default`
- if the system has several interfaces, `--interface` is not defined, and `default` does not exist, make `--interface` mandatory

The code is a bit ugly, but this might be the less intrusive change for users.

**Comment by SchoolGuy** [comment]
@btravouillon I like your approach. Is it possible that you create a testcase to cover this? It is logic that is important for users of the CLI after all...

**Comment by SchoolGuy** [comment]
@btravouillon You can drop the commit which fixes the rstcheck fixes and exchange that against this one: https://github.com/cobbler/cobbler/pull/3147/commits/d2778018e58f29456d344b4f0bd396f8285359a4 if you want. :)

**Comment by btravouillon** [comment]
I've been quite busy at work... Rebase is done to fix the issue with rstcheck.
I'm not a Python dev so I have no knowledge of pytest. Here is a first try, please tell me if that's ok so I can add other tests.

**Comment by SchoolGuy** [comment]
@btravouillon I have to check if the logic in regard to the interfaces is covered but on the phone this looks marvelous!

**Comment by btravouillon** [comment]
Not all the logic is covered yet. For now I only test if a system is created with a single interface named `default`. If ok, I will implement additional tests the same way.

**Comment by SchoolGuy** [comment]
@btravouillon Then go ahead. This test looks good.

**Comment by btravouillon** [comment]
Ready for review.

File: tests/xmlrpcapi/miscellaneous_test.py [hunk 1]
```diff
@@ -726,6 +726,309 @@ def test_xapi_object_edit(self, remote, token, remove_distro, create_kernel_init
...
+                {
+                    "name": name_system,
+                    "mac_address": "aa:bb:cc:dd:ee:ff",
+                },
+                token,
+            )
+        except (ValueError):
+            assert True
+        else:
+            pytest.fail("Missing interface name did not raise an exception!")
```
Comment (SchoolGuy) [inline tests/xmlrpcapi/miscellaneous_test.py:1030]: The pattern you used can be shortened to the following if I am not mistaken:

```suggestion
        with pytest.raises(ValueError):
            remote.xapi_object_edit(
                "system",
                name_system,
                "edit",
                {
                    "name": name_system,
                    "mac_address": "aa:bb:cc:dd:ee:ff",
                },
                token,
            )
```

**Review by SchoolGuy — APPROVED** [review]
LGTM with one minor suggestion which you can decide if you want to apply it or not.

File: tests/xmlrpcapi/miscellaneous_test.py (hunk 1 above)
Comment (btravouillon) [reply tests/xmlrpcapi/miscellaneous_test.py:1030]: That's fine.
The new error message might be less clear, but it is easy to find out what is wrong by reading the code.
```
| E           Failed: DID NOT RAISE <class 'ValueError'>
```

**Comment by SchoolGuy** [comment]
@btravouillon I just saw you didn't sign your commits. Is that possible for you or not?

**Pull request merged**